src/watchman_client.js

```diff
@@ -0,0 +1,492 @@
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ * 
+ * Note: WatchmanWatcher also added a 'watchmanPath' option for use with the sane CLI.
+ * Because of that, we actually need a map of watchman binary path to WatchmanClient instance.
+ * That is set up in getInstance(). Once the WatchmanWatcher has a given client, it doesn't
+ * change.
+ *
+ * @class WatchmanClient
+ * @param String watchmanBinaryPath 
+ * @public
+ */
+
+function WatchmanClient(watchmanBinaryPath) {
+
+  // define/clear some local state. The properties will be initialized
+  // in _handleClientAndCheck(). This is also called again in _onEnd when
+  // trying to reestablish connection to watchman.
+  this._clearLocalVars(); 
+
+  this._watchmanBinaryPath = watchmanBinaryPath;
+  this._clientListeners = null;  // direct listeners from here to watchman.Client.
+}
+
+// Define 'wildmatch' property, which must be available when we call the
+// WatchmanWatcher.createOptions() method. 
+Object.defineProperty(WatchmanClient.prototype, 'wildmatch', {
+  get() { return this._wildmatch; }
+});
+
+/**
+ * Called from WatchmanWatcher (or WatchmanClient during reconnect) to create
+ * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+ * watchman.Client, to be handled here.
+ */
+WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
+  let subscription;
+  let watcherInfo;
+  
+  return this._setupClient()
+    .then(() => {
+      watcherInfo = this._createWatcherInfo(watchmanWatcher);
+      subscription = watcherInfo.subscription;
+      return this._watch(subscription, root);
+    })
+    .then((resp) => this._clock(subscription))
+    .then((data) => this._subscribe(subscription))
+    .catch((error) => {
+      console.error("Failed creating subscription for root " + root);
+      throw error;
+    });
+};
+
+/**
+ * Remove the information about a specific WatchmanWatcher.
+ * Once done, if no watchers are left, clear the local vars,
+ * which will end the connection to the watchman.Client, too.
+ */
+WatchmanClient.prototype.closeWatcher = function(watchmanWatcher) {
+  let watcherInfos = Object.values(this._watcherMap);
+  let numWatchers = watcherInfos.length;
+
+  if (numWatchers > 0) {
+    let watcherInfo;
+    
+    for (let info of watcherInfos) {
+      if (info.watchmanWatcher === watchmanWatcher) {
+        watcherInfo = info;
+        break;
+      }
+    }
+
+    if (watcherInfo) {
+      delete this._watcherMap[watcherInfo.subscription];
+
+      numWatchers--;
+
+      if (numWatchers === 0) {
+        this._clearLocalVars();  // nobody watching, so shut the watchman.Client down.
+      }
+    }
+  }
+};
+
+/**
+ * Simple backoff-time iterator. next() returns times in ms.
+ * When it's at the last value, it stays there until reset()
+ * is called.
+ */
+WatchmanClient.prototype._backoffTimes = {
```

Comment:
As implemented, I believe state within `_backoffTimes` is shared between all WatchmanClients. The promises arises when more then one WatchmanClient instance is active (because we have multiple discrete watchman paths).

Given that it is possible to have multiple active WatchmanClient's, it is also possible for them both to be backing off at the same time. As implemented, the they will interfere which each other as they both share `_next: 0` state.

Reply:
Doh! Very good catch. I'd first come up with the backoffTimes thing before splitting for various paths. I'll move the _backoffTimes to an instance var (make a simple class first).